pkg/common/cns-lib/volume/listview.go

```diff
@@ -0,0 +1,348 @@
+	}
+	client.Timeout = noTimeout
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.govmomiClient = client
+	log.Infof("created new govmomi client after reload config")
+	return nil
+}
+
+// listenToTaskUpdates is a long-running goroutine
+// that uses a property collector to listen for task updates
+// CSI ops add CNS tasks to listview and wait for a response from CNS
+// when update(s) are received by the property collector,
+// it spawns a new goroutine to process each task update and return the result to the caller
+func (t *ListViewImpl) listenToTaskUpdates() {
+	log := logger.GetLogger(t.ctx)
+	filter := getListViewWaitFilter(t.listView)
+	// we need to recreate the listView and the wait filter after any error from vc
+	// for the first iteration we already have the listView and filter initialized
+	recreateView := false
+	for {
+		// calling Connect at the beginning to ensure the current session is neither nil nor NotAuthenticated
+		// if err := t.virtualCenter.Connect(t.ctx); err != nil {
+		if err := t.isClientValid(); err != nil {
+			log.Errorf("failed to connect to vCenter. err: %v", err)
+			time.Sleep(waitForUpdatesRetry)
+			continue
+		} else {
+			log.Infof("connection to vc successful")
+		}
+
+		if recreateView {
+			log.Info("re-creating the listView object")
+			err := t.createListView(t.ctx, nil)
+			if err != nil {
+				log.Errorf("failed to create a ListView. error: %+v", err)
+				// time.Sleep(waitForUpdatesRetry)
+				continue
+			}
+
+			filter = getListViewWaitFilter(t.listView)
+			recreateView = false
+		}
+
+		log.Info("Starting listening for task updates...")
+		t.mu.Lock()
+		pc := property.DefaultCollector(t.govmomiClient.Client)
+		t.mu.Unlock()
+		err := property.WaitForUpdates(t.ctx, pc, filter, func(updates []types.ObjectUpdate) bool {
+			log.Debugf("Got %d property collector update(s)", len(updates))
+			for _, update := range updates {
+				for _, prop := range update.ChangeSet {
+					log.Debugf("Got update for object %v properties %v", update.Obj, prop)
+					go t.processTaskUpdate(prop)
+				}
+			}
+
+			// this return value is used by the WaitForUpdates method.
+			// we only want this true while running the unit tests so the test can finish
+			return t.shouldStopListening
+		})
+		// if property collector returns any errors,
+		// we want to immediately return a fault for all the pending tasks in the map
+		// note: this is not a task error but an error from the vc
+		if err != nil {
+			log.Errorf("WaitForUpdates returned err: %v for vc: %+v", err, t.virtualCenter)
+			recreateView = true
+
+			t.mu.Lock()
+			for taskMoRef := range t.taskMap {
+				prop := types.PropertyChange{
+					Val: types.TaskInfo{
+						Task:  taskMoRef,
+						State: types.TaskInfoStateError,
+						Error: &types.LocalizedMethodFault{
+							LocalizedMessage: err.Error(),
+						},
+					},
+				}
+				go t.processTaskUpdate(prop)
```

Comment:
t.mu.Lock() is already locked outside, so there will be deadlock

Reply:
I've reworked the usage of locks with reader locks and writer locks as applicable. See https://go.dev/play/p/Gm22qn-lpq- for example.